### DISCUSSION TITLE
concrete_section.moment_interaction_diagram

### DISCUSSION THREAD

**Discussion by buddyd16** [robbievanleeuwen/concrete-properties#31] [General]
https://github.com/robbievanleeuwen/concrete-properties/blob/503a3f0692b685c1589618b74d4e874a58d1a232/concreteproperties/concrete_section.py#L940-L987

The docstring for the method notes it takes the neutral axis rotation as the entry parameter but does not appear to vary this value for each interaction point, need to be careful here. Holding the neutral axis rotation constant and solving for the different neutral axis depths to achieve the plot points does not actually give you the proper interaction diagram. If you plot the resulting points in 3D you'll see that they oscillate away from a constant Mx, My plane meaning the returned interaction points would not be in equilibrium with applied loading.

**Comment by robbievanleeuwen** [comment]
Hi @buddyd16, I see your point here, however the function does what it sets out to do. I don't think there's anything in the docs about keeping the moment ratio constant.

You are correct in noting that for bending about a non-principal axis the ratio between ``m_x`` and ``m_y`` will vary. The anlysis reflects this and these results are saved in the [``MomentInteractionResults``](https://robbievanleeuwen.github.io/concrete-properties/rst/results.html#moment-interaction-diagram) object, and the user can easily view these results. The ``plot_diagram()`` method plots either the ``m_x``, ``m_y`` or resultant component of the bending moment (``m_x`` by default).

From a design perspective I can see the value in requiring the ratio between moments is constant rather than the bending axis angle and I think this would be a useful feature to add! Do you have a reference on how this can be done? Seems like a much more difficult problem with two unknowns to be solved (as discussed on engtips forum)?

**Reply by buddyd16** [reply]
I misunderstood your plots, plotting only one component of the biaxial bending is unfortunate, that would provide a false design utilization ratio.

for design purposes you really need to plot the interaction curves on a constant mx, my plane consistent with the applied loads.

I have two spreadsheets for the problem in the open source group, the approach I took was to nest two root finding methods outer method for the neutral axis rotation and inner method for the neutral axis depth. A constrained optimization approach may be a better fit though minimizing sum of squares for the delta in axial load and deltas in i and j unit vectors for the moments with neutral axis depth constrained to 0 and a depth where the entire cross section is at concrete ultimate strain and the neutral axis angle constrained to be between -pi and pi.

**Reply by connorferster** [reply]
Maybe I am missing something here but it seems that the kind of plots @buddyd16 is looking for can be generated by the BiaxialBendingResults object where the results are plotted as mx vs my on a constant axial plane. Is this incorrect?

**Reply by buddyd16** [reply]
@connorferster @robbievanleeuwen hopefully this helps illustrate what I perceive to be an issue:
Cross Section:
![image](https://user-images.githubusercontent.com/12416202/183062824-27adee27-8686-47e0-9b0f-8c8389d97cea.png)

Loading:
![image](https://user-images.githubusercontent.com/12416202/183062929-e0633059-6a69-40b2-b7b9-05dd2e0839cb.png)

Mx vs My Slice at applied P:
![image](https://user-images.githubusercontent.com/12416202/183063233-a9c2ce6c-a201-40cc-9358-4fa9a063d402.png)

P v M at applied load angle atan2(mx,my):
![image](https://user-images.githubusercontent.com/12416202/183063504-a9d3c9de-01c2-40fa-9d9d-1abcde418475.png)

3D Interaction Surface showing the above diagrams and what it appears ConcreteProperties is doing:
![image](https://user-images.githubusercontent.com/12416202/183063952-bb08bfc1-7734-4e30-b73e-e873cfaf943f.png)

Top down view showing how a constant neutral angle plot strays from the true interaction curve:
![image](https://user-images.githubusercontent.com/12416202/183065151-5cd23915-9f4b-4718-bfb3-c520820c7cbc.png)

SPCol and other commercial apps are doing the P v M at applied load angle atan2(mx,my) line.

**Comment by robbievanleeuwen** [comment]
@buddyd16 thanks for the discussion and example, think this has finally clicked for me 😆. The current implementation works for bending about principal axes but as you describe does not give **useful** (read for design purposes - where ``m_x / m_y`` is known) interaction diagrams when there is biaxial bending.

This is showing how out of touch I have become with structural design after a few years out of the game! I've put my designer (rather than academic) hat back on and here's a bit of a brain dump - feel free to comment on this:

How users would currently interact with *concreteproperties*:
- Uniaxial bending about a principal plane ➡️ set ``theta`` equal to 0 or pi / 2 as required and generate interaction diagram - utilisation ratio can be obtained readily from this diagram as there is bending about one axis only ✅ 
- Biaxial bending present, user wants to generate an interaction diagram for a ratio of ``m_x`` to ``m_y`` ➡️ there is facility to do this in *concreteproperties* and user does not know which ``theta`` to set (as you describe there is no set ``theta`` for a given moment ratio!) - user is confused and cannot generate a moment interaction diagram for this case ❌ (worst case is ``theta=0`` is a non-principal bending axis and user takes these results as gospel ❌❌❌)
- User has a known axial force with biaxial bending ➡️ create biaxial bending diagram for known axial force ✅ (however a utilisation ratio cannot be obtained unless the user has zero axial force as axial force cannot be scaled on a constant axial force plot if it is non-zero ❌)
- **Case as described by @buddyd16:** User has biaxial bending with an unknown axial force and wants to find the utilisation ratio ➡️ only current option is to painstakingly generate a number of biaxial bending diagrams and somehow interpret these results, if only there was a way to create an interaction diagram for a constant ratio of ``m_x`` to ``m_y``!

**Solution:**
Add an option to generate a moment interaction diagram for a given ratio ``m_x / m_y`` rather than a netural axis angle ``theta`` (``theta`` is actually useless to the designer, but not to the programmer!) Thinking again about my original comment, would there not actually just be a single unknown that is ``theta``? At least in the way I have implemented my code I can vary the neutral axis depth from zero to infinity, and for each depth just solve for ``theta`` that gives the desired moment ratio? I don't think in generating the interaction diagram I would have two unknowns. I can see that in the general non-linear case that given a design axial force and moments, the neutral axis depth and neutral axis angle are unknown, thus the need for something more involved, but we don't have this problem do we? Keen to hear your thoughts on this!

**Issues with above solution:**
There is currently an option to specify control points in the moment interaction diagram, two of these would be quite difficult with this new implementation.
- Given axial force, adds another unknown and would involve constrained optimisation. I don't see how nested root finding would work as neutral axis depth and neutral axis angle aren't independent variables when it comes to calculating the net axial force (this is a coupled problem).
- Given strain in an extreme bar (used to find balanced point etc.), again adds another unknown to the problem.

Anyway that's a bit of an unstructured rant, hoping it adds to this discussion and makes at least some sense (I guess this is what happens when you try to write  a response five times and keep getting interrupted by your 6 week old daughter, how inconsiderate!) Keen to hear your thoughts and look forward to implementing a solution to this!

**Reply by buddyd16** [reply]
Go tend to that 6 week old, my kiddo turns two today enjoy these early times. I’ll dig in and get you a response to the above later this weekend.

**Reply by buddyd16** [reply]
Some Thoughts:
- Don't want to rely on m_x/m_y ratio as there are instances where m_y can be zero and the neutral axis angle, theta, would also not be 0. I landed on comparing the angle between the applied moments and calculated moments.

classic example of this is an L beam where textbooks have you hold the NA theta to 0, but this does not actually produce true uniaxial bending:
![image](https://user-images.githubusercontent.com/12416202/183454195-e42342e9-4191-4ea8-a68e-bc9e3e40402c.png)

True solution with only an Mx capacity value:
![image](https://user-images.githubusercontent.com/12416202/183454416-8bd07961-787c-4688-ae67-a5aab972aafe.png)


- Think there are multiple options for giving a designer a design utilization
-- Constant P, s * <m_x, m_y>, where s is the utilization scale factor and <m_x, m_y> is the 2D unit vector in the Constant P plane of the user applied loading.
-- s * <m_x, m_y, p>, where s is the utilization scale factor and <m_x, m_y, p> is the 3D unit vector of the user applied loading.
-- shortest distance from the user load point to the interaction surface constrained to the m_x, m_y plane.
![image](https://user-images.githubusercontent.com/12416202/183455166-5ccf4459-54fe-4ac8-858f-524f3c62b3c3.png)


> At least in the way I have implemented my code I can vary the neutral axis depth from zero to infinity, and for each depth just solve for theta that gives the desired moment ratio?

The problem you'll run into here is that the neutral axis depth required to achieve the user applied P is dependent on the neutral axis theta. The L beam is an example where working through a list of neutral axis depths may yield to sparse results for the interaction curve.

> Given axial force, adds another unknown and would involve constrained optimisation. I don't see how nested root finding would work as neutral axis depth and neutral axis angle aren't independent variables when it comes to calculating the net axial force (this is a coupled problem).

nested roots work since the neutral axis depth for P is inherently tied to the neutral axis angle theta. 
here are two spreadsheets one covering T,L and rectangular section and one covering circles:
https://github.com/open-struct-engineer/Excel/blob/master/03-Concrete/03_T_L_R_Beam%20Capacity_MR.xlsm
https://github.com/open-struct-engineer/Excel/blob/master/03-Concrete/03_Circular_Section_Analysis_MR.xlsm

> Given strain in an extreme bar (used to find balanced point etc.), again adds another unknown to the problem.

With a given strain you can solve for the required neutral axis depth directly from the linear strain equation so the only unknown would the neutral axis angle.




In my study of the problem there also exists conditions where the m_x v m_y plot in either the tension region or the upper end of the compression region actually shifts to exist in two or in extreme cases one quadrant of the m_x v m_y axis. In these cases root finding or optimization based on the applied load angle to resultant angle will fail and an alternate solution method is required. I found that a vector based approach yielded good results and also works well for cases where two solutions exist. See this eng-tips post on the subject: https://www.eng-tips.com/viewthread.cfm?qid=475741
( this is implemented as the "experimental" solution in the T,L,R spreadsheet)

* The coding in those spreadsheets is pretty garbage so if you need me to clarify any of it just let me know.

**Reply by robbievanleeuwen** [reply]
Thanks for the thoughts @buddyd16, unfortunately I don't have time to explore the biaxial stuff at the moment but hope to jump into it in the near future.

A few things still don't feel quite resolved for me yet, but they may become more apparent once I start to try to tackle the problem. E.g.

> Don't want to rely on m_x/m_y ratio as there are instances where m_y can be zero and the neutral axis angle, theta, would also not be 0. I landed on comparing the angle between the applied moments and calculated moments.

I think you might have missed my idea here, I would use the ``brentq`` solver to find ``theta`` such that the desired ``m_x/m_y`` ratio is obtained. So in the example of the L-beam, if the user desired for ``m_y`` to be zero, for each neutral axis depth in the interaction diagram, ``theta`` would be found to satisfy this moment ratio requirement (``theta`` would be non-zero as you correctly describe).

> With a given strain you can solve for the required neutral axis depth directly from the linear strain equation so the only unknown would the neutral axis angle.

If the neutral axis angle ``theta`` is unknown, the neutral axis depth cannot be found. This would require some more thought but I feel like there's an answer in there somewhere.

Anyway, I'll be putting this on ice until I have some proper time to dedicate to this. However, I greatly value your experience on this one @buddyd16 and when I get back into it I hope to use these ideas to arrive at a good solution!

**Reply by buddyd16** [reply]
> Thanks for the thoughts @buddyd16, unfortunately I don't have time to explore the biaxial stuff at the moment but hope to jump into it in the near future.

Yeah I've been stopped dead in my tracks on my little side programming endeavor, this is a great tool you've put together so far excited to see how it matures. I'll be peaking into the code when I can and hopefully offer more than just this discussion text. 
 
> A few things still don't feel quite resolved for me yet, but they may become more apparent once I start to try to tackle the problem. E.g.

> > With a given strain you can solve for the required neutral axis depth directly from the linear strain equation so the only unknown would the neutral axis angle.
 
> If the neutral axis angle `theta` is unknown, the neutral axis depth cannot be found. This would require some more thought but I feel like there's an answer in there somewhere.

Speaking more towards the root finding process, if you go the option of nesting two brentq solvers you would pass a neutral axis and an optional required extreme tensile strain into the outer solver, if the strain was passed in solve for neutral axis depth directly from the linear strain equation if no strain passed in then proceed to the inner brentq solver for the neutral axis depth. 

> Anyway, I'll be putting this on ice until I have some proper time to dedicate to this. However, I greatly value your experience on this one @buddyd16 and when I get back into it I hope to use these ideas to arrive at a good solution!

Looking forward to being able to help out. You've made incredibly strides on the state of the tool already.

**Reply by Agent6-6-6** [reply]
To add to the discussion, until the analysis based on load angle is implemented the only true way to see how close you are to the curve is to compare on a biaxial plot.

To show the effect of the 'wander' of the interaction surface based on the constant neutral axis angle, you can see the effect at the base of this interaction surface created by concrete-properties (look at the black lines). The section was a 600 square column with 12-DH20 bars.

![3D_plot](https://user-images.githubusercontent.com/36808164/190832822-56aea06a-f3ff-4443-b5df-3d21e5c4626a.svg)

Where you can see the vertical black lines all tend to converge at 4 specific orientations, or further up they wander around the surface which means a lot less accuracy as you get near to the surface point especially if you're relying on the M/N curves and a load point lying inside this. You really should only be comparing $M_x$ and $M_y$ load data on the biaxial plane as this is correct

**Reply by robbievanleeuwen** [reply]
Hi @Agent6-6-6, cool plot! Would you mind sharing the code you used to generate this? It will be helpful for when I get to tackling this one 😄

**Reply by Agent6-6-6** [reply]
@robbievanleeuwen, sure something along these lines was used.

Just sub in the max compression/tension loads and strength reduction factors from the AS3600 code if using that (as the example code was generated based on #40). Note `design_code=NZS3101()` in the example.

Might be nice to have a standard method to generate something like this? 

Note takes a while to run though!


```python
# determine maximum compression load for the column section
phi = 0.85
max_comp = phi * design_code.max_comp_strength()
max_ten = phi * design_code.max_ten_strength()
steps = 15
# generate axial load list
n_list = np.linspace(-max_ten + 20000, max_comp, steps)
theta_list = np.linspace(0.0, 2 * np.pi, 48, False)

results_bb = []
results_mi = []

for theta in theta_list:
    f_mi_res, _, _ = design_code.moment_interaction_diagram(theta=theta)
    results_mi.append(f_mi_res)
for n in n_list:
    f_bb_res, _ = design_code.biaxial_bending_diagram(n=n)
    results_bb.append(f_bb_res)

# plot all the diagrams on one image
fig = plt.figure(figsize=(12, 10))
ax = plt.axes(projection="3d")
n_scale = 1e-3
m_scale = 1e-6
for mi_res in results_mi:
    n, m_x_list = mi_res.get_results_lists(moment="m_x")
    _, m_y_list = mi_res.get_results_lists(moment="m_y")
    n = np.array(n) * n_scale
    m_x_list = np.array(m_x_list) * m_scale
    m_y_list = np.array(m_y_list) * m_scale

    ax.plot3D(m_x_list, m_y_list, n, "-k", linewidth=0.5)

for i, bb_res in enumerate(results_bb):
    m_x_list, m_y_list = bb_res.get_results_lists()
    m_x_list = np.array(m_x_list) * m_scale
    m_y_list = np.array(m_y_list) * m_scale

    n = np.array(n_list[i]) * n_scale
    ax.plot3D(m_x_list, m_y_list, n, "-r", linewidth=0.5)

ax.set_xlabel("Bending Moment $M_x$")
ax.set_ylabel("Bending Moment $M_y$")
ax.set_zlabel("Axial Force $N$")

plt.show()
```
